nexus/src/app/certificate.rs

```diff
@@ -0,0 +1,230 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! x.509 Certificates
+
+use crate::context::OpContext;
+use crate::db;
+use crate::db::lookup;
+use crate::db::lookup::LookupPath;
+use crate::db::model::Name;
+use crate::external_api::params;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::NameOrId;
+use openssl::pkey::PKey;
+use openssl::x509::X509;
+use ref_cast::RefCast;
+use std::sync::Arc;
+use uuid::Uuid;
+
+fn validate_certs(input: Vec<u8>) -> Result<(), String> {
+    let certs = X509::stack_from_pem(&input.as_slice())
+        .map_err(|err| format!("Failed to parse certificate as PEM: {err}"))?;
+    if certs.is_empty() {
+        return Err("could not parse".to_string());
+    }
+    Ok(())
+}
+
+fn validate_private_key(key: Vec<u8>) -> Result<(), String> {
+    let _ = PKey::private_key_from_pem(&key.as_slice())
+        .map_err(|err| format!("Failed to parse private key as PEM: {err}"))?;
+
+    Ok(())
+}
+
+impl super::Nexus {
+    pub fn certificate_lookup<'a>(
+        &'a self,
+        opctx: &'a OpContext,
+        certificate: &'a NameOrId,
+    ) -> lookup::Certificate<'a> {
+        match certificate {
+            NameOrId::Id(id) => {
+                LookupPath::new(opctx, &self.db_datastore).certificate_id(*id)
+            }
+            NameOrId::Name(name) => LookupPath::new(opctx, &self.db_datastore)
+                .certificate_name(Name::ref_cast(name)),
+        }
+    }
+
+    pub async fn certificate_create(
+        self: &Arc<Self>,
+        opctx: &OpContext,
+        params: params::CertificateCreate,
+    ) -> CreateResult<db::model::Certificate> {
+        validate_certs(params.cert.clone()).map_err(|e| {
+            warn!(self.log, "bad cert: {e}");
+            Error::InvalidValue { label: String::from("cert"), message: e }
+        })?;
+        validate_private_key(params.key.clone()).map_err(|e| {
+            warn!(self.log, "bad key: {e}");
+            Error::InvalidValue { label: String::from("key"), message: e }
+        })?;
+
+        let new_certificate = db::model::Certificate::new(
+            Uuid::new_v4(),
+            db::model::ServiceKind::Nexus,
+            params,
+        );
+
+        // TODO: Saga?
```

Comment:
Why?  (when would we have enough information to resolve this TODO?)

Reply:
Updated in 400d90c8528b936c873df2c9ca317344c079a542

Reply:
Thanks!